File: apps/go-api/internal/store/observe.go
package store

import (
	"context"
	"time"

	"prism/go-api/internal/metrics"
	"prism/go-api/internal/model"
)

type Observed struct {
	inner Store
}

func Observe(inner Store) Store {
	return &Observed{inner: inner}
}

func (s *Observed) Name() string { return s.inner.Name() }

func (s *Observed) Close() error { return s.inner.Close() }

func (s *Observed) Ping(ctx context.Context) error {
	start := time.Now()
	err := s.inner.Ping(ctx)
	metrics.ObserveStorage(s.Name(), "ping", time.Since(start), err)
	return err
}

func (s *Observed) Write(ctx context.Context, samples []model.Sample) error {
	start := time.Now()
	err := s.inner.Write(ctx, samples)
	metrics.ObserveStorage(s.Name(), "write", time.Since(start), err)
	return err
}

func (s *Observed) Locf(ctx context.Context, tagIDs []uint32, at time.Time) ([]model.Sample, error) {
	start := time.Now()
	res, err := s.inner.Locf(ctx, tagIDs, at)
	metrics.ObserveStorage(s.Name(), "locf", time.Since(start), err)
	return res, err
}

func (s *Observed) Range(ctx context.Context, tagIDs []uint32, from, to time.Time) ([]model.Sample, error) {
	start := time.Now()
	res, err := s.inner.Range(ctx, tagIDs, from, to)
	metrics.ObserveStorage(s.Name(), "range", time.Since(start), err)
	return res, err
}

func (s *Observed) UpsertTags(ctx context.Context, tags []model.Tag) error {
	start := time.Now()
	err := s.inner.UpsertTags(ctx, tags)
	metrics.ObserveStorage(s.Name(), "tags", time.Since(start), err)
	return err
}

func (s *Observed) ListTags(ctx context.Context) ([]model.Tag, error) {
	start := time.Now()
	res, err := s.inner.ListTags(ctx)
	metrics.ObserveStorage(s.Name(), "tags", time.Since(start), err)
	return res, err
}

File: apps/python-api/app/store/observe.py
from datetime import datetime

from app.metrics import observe_storage, track
from app.models import Sample, Tag
from app.store.base import Store


class ObservedStore:
    def __init__(self, inner: Store) -> None:
        self._inner = inner
        self.name = inner.name

    async def ping(self) -> None:
        with track() as elapsed:
            try:
                await self._inner.ping()
            except Exception as exc:
                observe_storage(self.name, "ping", elapsed(), exc)
                raise
        observe_storage(self.name, "ping", elapsed())

    async def write(self, samples: list[Sample]) -> None:
        with track() as elapsed:
            try:
                await self._inner.write(samples)
            except Exception as exc:
                observe_storage(self.name, "write", elapsed(), exc)
                raise
        observe_storage(self.name, "write", elapsed())

    async def locf(self, tag_ids: list[int], at: datetime) -> list[Sample]:
        with track() as elapsed:
            try:
                result = await self._inner.locf(tag_ids, at)
            except Exception as exc:
                observe_storage(self.name, "locf", elapsed(), exc)
                raise
        observe_storage(self.name, "locf", elapsed())
        return result

    async def range(self, tag_ids: list[int], start: datetime, end: datetime) -> list[Sample]:
        with track() as elapsed:
            try:
                result = await self._inner.range(tag_ids, start, end)
            except Exception as exc:
                observe_storage(self.name, "range", elapsed(), exc)
                raise
        observe_storage(self.name, "range", elapsed())
        return result

    async def upsert_tags(self, tags: list[Tag]) -> None:
        with track() as elapsed:
            try:
                await self._inner.upsert_tags(tags)
            except Exception as exc:
                observe_storage(self.name, "tags", elapsed(), exc)
                raise
        observe_storage(self.name, "tags", elapsed())

    async def list_tags(self) -> list[Tag]:
        with track() as elapsed:
            try:
                result = await self._inner.list_tags()
            except Exception as exc:
                observe_storage(self.name, "tags", elapsed(), exc)
                raise
        observe_storage(self.name, "tags", elapsed())
        return result

    async def close(self) -> None:
        await self._inner.close()
